content/blog/2020-10-28-type-based-alias-analysis.md

```diff
@@ -0,0 +1,138 @@
+would have two different, incompatible types. _FieldTypeDecl_ takes this 
+further by adding checks for field names, array accesses, and pointer 
+dereferences. For example, paths `p.f` and `q.g` may be aliases only if the 
+field names `f` and `g` are exactly the same and `p` and `q` themselves may be 
+aliases. Additionally, a pointer dereference `q^` may alias a field `p.f` or an 
+array access `p[i]` only if its address was ever taken by the program (since 
+otherwise no pointer to it could exist). Further checks exist for comparisons 
+between pointers and array accesses, which the authors sum up as follows:
+
+1.	Identical [access paths] always alias each other.
+2.	Two qualified expressions may be aliases if they access the same field in 
+potentially the same object.
+3.	A pointer dereference may reference the same location as a qualified or 
+subscripted expression only if their types are compatible and the program may 
+take the address of the qualified or subscripted expression.
+4.	See (3)
+5.	In Modula-3, a subscripted expression cannot alias a qualified expression.
+6.	Two subscripted expressions are aliases if they may subscript the same 
+array. _FieldTypeDecl_ ignores the actual subscripts.
+7.	For all other cases of [access paths], including two pointer dereferences,
+_FieldTypeDecl_ uses _TypeDecl_ to determine aliases.
+Finally, _SMFieldTypeRefs_ uses a simple flow-insensitive pass through the 
+program’s instructions to merge types that might become aliases. It uses a 
+table that tracks the equivalence classes for a given type (such that for a 
+type `T` with entry `{T1, T2, …}`, an access path of type `T` may be a 
+reference to any `T1, T2, …`). Upon encountering an assignment from type `T1` 
+to a pointer of type `T2`, it adds `T1`’s entry to `T2`’s equivalence class. 
+Notably, this is an asymmetric operation; the assignment does not change `T1`’s 
+equivalence class. This is a more refined version of _TypeDecl_ above, so by 
+replacing _TypeDecl_ with lookups from this table (called _SMTypeRefs_ in the 
+paper, for “Selectively Merge Type References”), we have the final algorithm 
+_SMFieldTypeRefs_. 
+
+## Evaluation
+The paper does a surprisingly thorough evaluation job considering the paper 
+was written in 1998. The authors observe that previous alias analyses were 
+evaluated by their static properties, like the size of the set of alias pairs, 
+and their dynamic properties, such as the impact of an alias analysis on a 
+compiler optimization that uses it. The static evaluation, performed over a 
+selection of 10 benchmarks, finds that _TypeDecl_ performs much worse than 
+_FieldTypeDecl_ in terms of alias set size, but _SMFieldTypeRefs_ barely 
+improves on _FieldTypeDecl_ at all. It also shows that applying type-based 
+alias analysis interprocedurally generates huge numbers of aliases, rendering 
+this type of analysis infeasible for global optimization. 

```

Comment:
Do you mean interprocedural or global? (It can't be both.)

Reply:
whoops, confused global with program wide